Mil sym support line types

youngca opened the pull request:
See corresponding app branch milSymSupportLineTypes.

youngca committed:
Add Follow and Support & Follow and Assume graphics.

youngca committed:
Add simplify library for optimizing the number of line vertices - removes redundant points.

This normally would have been an npm module, but require script errors kept firing, so this has been
reconfigured from the original node module.
Special functions to handle Konva-formatted point arrays and mil-sym point strings were added.

youngca committed:
Reduce the number of vertices published in the state for free-draw and polygon objects.

Depending upon the shape of the free-hand drawing or polygon, there can be a substantial reduction
in the number of vertices, thus reducing the size of the Konva object in the state.

youngca committed:
Add definitions for 23 new line and wire/fence type of mission graphics.

youngca committed:
Code cleanup.

youngca committed:
Add acceptable modifiers.

youngca committed:
Add tactical graphic-specific modifiers.

youngca committed:
Mission graphics render - return image object instead of just dataURL so we can get image attributes.

youngca committed:
Provide a method to get the computed cornerpoints of the graphic from the mission graphic image.

youngca committed:
Add unique designations to supply route graphics.

youngca committed:
Add multipointPolygon types of mission graphics.

youngca committed:
Update mil-sym to rebase of missioncommand's version 0.3.29.

youngca committed:
Store graphic bounds as its own property, separate from Konva Image width and height.

mil-sym uses the bounds to produce an image.  The image dimensions don't always conform to the
extents defined in the bounds because of allowances for glyphs and modifiers, but needs to be saved
to produce a consistently-image regardless of modifiers.

youngca committed:
Map the "ModifiersTG" alias to armyc2.c2sd.renderer.utilities.ModifiersTG.

youngca committed:
Don't bother simplifying if there are not at least 2 points in the array.

eric79 commented on support/client/lib/vwf/model/mil-sym.js:
```diff
@@ -450,22 +450,29 @@ define( [ "module",
-                                node[ propertyName ] = propertyValue;
-                                renderImage = basicPropertiesMet( node );
+                        case "bounds":
+                            if ( node.nodeType === "missionGfx" && propertyValue !== undefined &  propertyValue !== null ) {
```
It looks like you're missing the second `&` on that last conditional.

eric79 replied:
It looks like this comment still needs to be addressed.

eric79 commented on support/client/lib/vwf/model/mil-sym.js:
```diff
@@ -450,22 +450,29 @@ define( [ "module",
+                                //var attrsChanged = ( attrs.width !== node.width) || ( attrs.height !== node.height);
+                                //if ( attrsChanged ) {
+                                    //node.width = attrs.width;
+                                    //node.height= attrs.height;
```
Can the commented-out code go away?

eric79 commented on support/client/lib/vwf/model/mil-sym.js:
```diff
@@ -700,10 +707,10 @@ define( [ "module",
-            var img = rendererMP.RenderSymbol2D(node.ID,node.fullName,node.description, symbolCode, controlPts, node.width, node.height, null, node.modifiers, format);
+            var img = rendererMP.RenderSymbol2D(node.ID,node.fullName,node.description, symbolCode, controlPts, node.bounds[0], node.bounds[1], null, node.modifiers, format);
 
             if ( !!img && !!img.image ) {
```
I've become a fan of @davideaster 's way of chaining non-null checks like this:

    if ( ( img || {} ).image ) {

... I offer it in case you find it interesting.  :smile:

eric79 commented on support/client/lib/vwf/model/mil-sym.js:
```diff
@@ -783,6 +796,10 @@ define( [ "module",
                     modifierActualName = msa[ modObj.modifier ];
                     break;
                     
+                case "ModifiersTG":
```
It looks like this `switch` statement is a candidate for being factored out into its own function since it is repeated.

eric79 commented on support/client/lib/vwf/utility/vertexSimplify/simplify.js:
```diff
@@ -0,0 +1,188 @@
+        },
+
+        // simplification using optimized Douglas-Peucker algorithm with recursion elimination
+        simplifyDouglasPeucker: function(points, sqTolerance) {
```
I bet Douglas Peucker got made fun of quite a bit as a child....

youngca replied:
Probably had a hard time getting dates.  :-)

eric79 commented on support/client/lib/vwf/utility/vertexSimplify/simplify.js:
```diff
@@ -0,0 +1,188 @@
+
+        // Simplify konva points
+        // Konva points are in an array like [x1,y1,x2,y2,...xn,yn]
+        simplifyKonvaPts: function( points ) {
```
Did you add these functions, I assume?  What would you think about moving them out of the library code into the appropriate view drivers, such that they could be invoked from our app like `setViewProperty`, etc?  That way we wouldn't have to modify the third-party code.

youngca replied:
I already had to modify the third-party code so we could load it, but I see your point.  I'll see what I can do.

eric79 commented on support/client/lib/vwf/view/mil-sym.js:
```diff
@@ -368,25 +369,39 @@ define( [ "module", "vwf/view", "mil-sym/cws", "jquery" ], function( module, vie
-        }      
+        if ( !!img ) {
+            return img;
+        }    
```
Since, if `img` is falsy, this function will return `undefined`, can you just `return img;` without the condition?

eric79 commented on support/client/lib/vwf/view/mil-sym.js:
```diff
@@ -368,25 +369,39 @@ define( [ "module", "vwf/view", "mil-sym/cws", "jquery" ], function( module, vie
         return;
     }
 
+    function convertKonvaToMilSymControlPts( konvaControlPts ) {
```
Let's remove the reference to `konva` inside the `mil-sym` driver.  Ideally, `mil-sym` should not know that `konva` exists.

Since this converts an array into a single value, it is a candidate for using the js array `reduce` function.  You could probably inline that in the code above, and avoid the reference to `konva` entirely.

eric79 reviewed (changes requested):
I've marked this as "Request changes" mostly because of the bugs that I discovered and commented on in your app PR.

youngca committed:
Remove redundant mission graphic.

youngca committed:
In image.onload, only update the node width and height if the image dimensions have actually changed.

The node width and height for a Konva image object determines the scaling of the image.

youngca committed:
Address refactoring and other issues and suggestions from PR.

- Chain non-null checks the Easter way in model/mil-sym.js
- Refactor switch statement for modifierActualName into its own functions
- Return img whether defined or not
- Remove references to konva in mil-sym driver

youngca committed:
Move KonvaPts simplify function to kinetic driver.

eric79 commented on support/client/lib/vwf/model/kineticjs.js:
```diff
@@ -2153,8 +2153,11 @@ define( [ "module",
+                if ( ( kImg.width !== imageObj.width ) || ( kImg.height !== imageObj.height ) ) {
+                    kineticObj.width( imageObj.width );
+                    kineticObj.height( imageObj.height );
+                }
```
It looks like it might make sense to split these into two conditionals:  Update the width if the width changed, and update the height if the height changed.

eric79 commented on support/client/lib/vwf/model/mil-sym.js:
```diff
@@ -452,36 +452,29 @@ define( [ "module",
-                                    renderImage = basicPropertiesMet( node );
-                                }
+                                node[ propertyName ] = propertyValue;
+                                renderImage = basicPropertiesMet( node );
```
I'm curious ... why did this need to change to set the value even if it's the same as what already existed?

eric79 commented on support/client/lib/vwf/model/mil-sym.js:
```diff
@@ -452,36 +452,29 @@ define( [ "module",
-                    var msa = armyc2.c2sd.renderer.utilities.MilStdAttributes;
-                    var mu = armyc2.c2sd.renderer.utilities.ModifiersUnits;
+                    //var msa = armyc2.c2sd.renderer.utilities.MilStdAttributes;
+                    //var mu = armyc2.c2sd.renderer.utilities.ModifiersUnits;
```
Can these commented-out lines go away?

eric79 commented on support/client/lib/vwf/model/mil-sym.js:
```diff
@@ -696,7 +689,7 @@ define( [ "module",
             var scale = 100.0;
             var renderer = armyc2.c2sd.renderer;
-            var msa = renderer.utilities.MilStdAttributes;
+            //var msa = renderer.utilities.MilStdAttributes;
```
Can this line go away?

eric79 commented on support/client/lib/vwf/model/mil-sym.js:
```diff
@@ -709,44 +702,51 @@ define( [ "module",
             var img = rendererMP.RenderSymbol2D(node.ID,node.fullName,node.description, symbolCode, controlPts, node.bounds[0], node.bounds[1], null, node.modifiers, format);
 
-            if ( !!img && !!img.image ) {
+            //if ( !!img && !!img.image ) {
```
Can this line go away?

eric79 commented on support/client/lib/vwf/model/mil-sym.js:
```diff
@@ -709,44 +702,51 @@ define( [ "module",
-                default:
-                    modelDriver.logger.errorx( "setModifier", "Unknown type (", modObj.type, ") specified." );
-                    return modifierSet;
+            var modifierActualName = getModifierActualName( modObj );
```
I like it.  :+1:

eric79 commented on support/client/lib/vwf/view/kineticjs.js:
```diff
@@ -1283,6 +1284,33 @@ define( [ "module", "vwf/view", "jquery", "vwf/utility", "vwf/utility/color", "v
+        var ptarray = [];
+        // Convert to array of  x, y points
+        for ( var i = 0; i < points.length; i = i+2 ) {
+            var point = { x: points[i], y: points[i+1] };
```
Just a reminder that we put spaces inside `[]`.

eric79 commented on support/client/lib/vwf/view/mil-sym.js:
```diff
@@ -374,27 +374,29 @@ define( [ "module", "vwf/view", "mil-sym/cws", "jquery" ], function( module, vie
-        }    
+        //if ( !!img ) {
+        //    return img;
+        //}    
```
Can this go away?

eric79 reviewed (changes requested):
This is looking good.  I'm marking it as "Request changes" mostly for the single `&` in that one conditional.

youngca committed:
Respond to PR comments.

- Address whitespace issues
- Remove commented out code
- Fix single & in conditional
- Set width and height changes in separate conditionals
